samza-core/src/main/java/org/apache/samza/storage/StoreBackupEngine.java

```diff
@@ -0,0 +1,185 @@
+    if(!dbDir.exists()) {
+      dbDir.mkdirs();
+    }
+
+    // Variables to keep track of statistics
+    long startTime = System.currentTimeMillis();
+    long totalBytesCopied = 0;
+    long totalBytesDeleted = 0;
+
+    // Find what files to copy from backup compare to what's in Rocksdb
+    ArrayList<FileStatus> backupFilesToCopy = getNewFilesInBackupToLocal(dbPath);
+    for (FileStatus backupFile : backupFilesToCopy) {
+      LOGGER.info("Found file: " + backupFile.getPath().getName());
+      totalBytesCopied += backupFile.getLen();
+      // Method: copyToLocalFile(hdfsFilePath, localFilePath)
+      fs.copyToLocalFile(backupFile.getPath(), dbPath);
+    }
+
+    metrics.bytesRestored().inc(totalBytesCopied);
+    metrics.numFilesRestored().inc(backupFilesToCopy.size());
+    // Print statistics of restoring a backup
+    long endTime = System.currentTimeMillis();
+    long timeElapsed = endTime - startTime;
+    LOGGER.info("Restoring RocksDB finished in milliseconds: " + timeElapsed);
+    LOGGER.info("Copied " + backupFilesToCopy.size() + " files to rocksdb directory");
+    LOGGER.info("Total size copied over: " + totalBytesCopied);
+
+    // After delete the extra files in db that aren't in backup
+    ArrayList<File> dbFilesToDelete = getNewFilesInLocalToBackup(dbPath);
+    for (File dbFile : dbFilesToDelete) {
+      LOGGER.info("Found file to delete: " + dbFile.getName());
+      totalBytesDeleted += dbFile.length();
+      dbFile.delete();
+    }
+
+    // Print statistics
+    metrics.numFilesDeleted().inc(dbFilesToDelete.size());
+    metrics.bytesDeleted().inc(totalBytesDeleted);
+    LOGGER.info("Deleted " + dbFilesToDelete.size() + " files from rocksdb directory");
+    LOGGER.info("Total size deleted: " + totalBytesDeleted);
+
+    return true;
+  }
+
+
+  // Method to find new files in backup compared to local
+  public ArrayList getNewFilesInBackupToLocal(Path localPath) throws IOException {
+    HashMap<String, File> localFileHmap = new HashMap<String, File>();
+    ArrayList<FileStatus> filesNotInLocalList = new ArrayList<FileStatus>();
+    // Store all the Rocksdb files in hashmap for quick comparison
+    File dbDir = new File(localPath.toString());
+    File[] filesList = dbDir.listFiles();
+    for (File dbFile : filesList) {
+      String fileName = dbFile.getName();
+      localFileHmap.put(fileName, dbFile);
+    }
+
+    // Get all files in the hdfs backup and compare to local. Add if different/doesn't exist
+    FileStatus[] backupFileIterator = fs.listStatus(hdfsPath);
+    for (FileStatus backupFile : backupFileIterator) {
+      String fileName = backupFile.getPath().getName();
+
+      // If found locally, check filesize to see if needed
+      if (localFileHmap.containsKey(fileName)) {
+        // TODO: This may not be good enough check if things fail to copy or weird things happen, probably an edge case here to consider
+        if (backupFile.getLen() > localFileHmap.get(fileName).length()) {
```

Comment:
This should be an error and we should either fail, or log and overwrite. Preferably the former for now so that we can investigate how this hapenned.